### DISCUSSION TITLE
Suggestion on page 441

### DISCUSSION THREAD

**Discussion by baifanhorst** [rasbt/machine-learning-book#124] [General]
In the definition of MultiLayerPerceptron on page 441, a function called 'training_epoch_end' is defined. However, this causes the following error:

NotImplementedError: Support for `training_epoch_end` has been removed in v2.0.0. `MultiLayerPerceptron` implements this method. You can use the `on_train_epoch_end` hook instead. To access outputs, save them in-memory as instance attributes. You can find migration examples in https://github.com/Lightning-AI/lightning/pull/16520.

According to the error message, we need to replace the definition of 'training_epoch_end' by the following:

def on_train_epoch_end(self):  
    self.log("train_acc", self.accuracy_train.compute())

**Comment by JevZin** [comment]
Thank you for this suggestion! There's just one small mistake: "accuracy_train" was never defined. Instead, the second line should use "train_acc" as follows:

self.log("train_acc", self.train_acc.compute())

**Comment by rasbt** [comment]
Sorry for the late response and thanks for jumping in here explaining the issue. I am just catching  up with my GitHub inbox and uploaded a fix here: #194
